Build guide: Calling zome functions (#528)
holochain/docs-pages: src/pages/build/calling-zome-functions.md

---
title: "Calling Zome Functions"
---

::: intro
You can call a zome function from [various local and remote processes](/build/connecting-the-parts/#what-processes-can-connect-to-a-happ) and secure them with capability tokens, letting you implement secure interop easily.
:::

## Call a zome function from a front end

To call a zome function, use [`AppWebsocket.prototype.callZome`](https://github.com/holochain/holochain-client-js/blob/main/docs/client.appwebsocket.callzome.md). It takes a [`CallZomeRequest`](https://github.com/holochain/holochain-client-js/blob/main/docs/client.callzomerequest.md) and an optional timeout.

> This example gets all the movies by a given director, then logs their titles to the console. You'll need the [`getHolochainClient` example](/build/connecting-a-front-end/#connect-to-a-happ-with-the-javascript-client) from the Connecting to a Front End page.

Reviewer: Which "Connecting to a Front End page"? Can you link to it?

Author: oh, the link is there in the paragraph -- I'll make it wrap all the pertinent text.

```typescript
import { decodeHashFromBase64, EntryHash } from '@holochain/client';

> class Movie {

Reviewer: I'd go with an `interface` instead of a `class`. I think that's more idiomatic, but I don't have evidence for it =)

    title: string;
    director_hash: EntryHash;
    imdb_id: string | null;
    release_date: Date;
    box_office_revenue: Number;
}

async function logMoviesBySergioLeone() {
    // Use the `getHolochainClient` function from /build/connecting-a-front-end/
    const client = await getHolochainClient();
    const movies: Array<Movie> = await client.callZome({
        role_name: "movies",
        zome_name: "movies",
        fn_name: "get_movies_for_director",
        // This should match the input payload struct for the zome function.
        // The client serializes and deserializes hashes as 39-byte
        // `Uint8Array`s.
        // If your client ever needs to work with hashes as strings, the
        // helper functions `decodeHashFromBase64` and `encodeHashToBase64`
        // will do the conversions for you.
        payload: decodeHashFromBase64(" -- base64-encoded Sergio Leone hash -- ")
    });
    console.log("Got movies");
    for (let movie in movies) {
        console.log(movie.title);
    }
}
```

> The client handles errors (both `ExternResult::Err(_)` errors from the zome function and other errors from Holochain itself) and deserializes the MessagePack return value for you. It also takes care of capability security<!--TODO: link when written --> for you.

Reviewer 2: > It also takes care of capability security

Is this referring to the client? What is meant exactly by "taking care of capability security" here, that it has the zome call signed by launcher/kangaroo/... with the author key under the hood?

Author: Right. I'll try to make that clearer without going into too much detail.

Reviewer 2: great, looking clearer now :+1:

## Call a zome function from another zome in the same hApp

> Coordinator zomes can't call each other's functions directly; instead they do it via the [`hdk::prelude::call`](https://docs.rs/hdk/latest/hdk/p2p/fn.call.html) host function.

Reviewer: That looks direct to me.
```suggestion
Coordinator zomes can call each other's functions directly via the [`hdk::prelude::call`](https://docs.rs/hdk/latest/hdk/p2p/fn.call.html) host function.
```

This example calls a function `get_movies_by_director` in another zome called `movies` _in the same cell_:

```rust

> fn get_movies_by_sergio_leone() -> ExternResult<Vec<Movie>> {
>     let response = call(
>         // This indicates that `bar` is in the same cell.
>         CallTargetCell::Local,
>         "movies",
>         "get_movies_for_director".into(),
>         // This is a capability secret -- we'll explain later why it's not
>         // needed.
>         None,
>         EntryHash::from_raw_36(vec![/* bytes of Sergio Leone director entry's hash */]),
>     )?;

Reviewer: ```suggestion
fn get_movies_for_director(director: EntryHash) -> ExternResult<Vec<Movie>> {
    let response = call(
        // This indicates that `bar` is in the same cell.
        CallTargetCell::Local,
        "movies",
        "get_movies_for_director".into(),
        // This is a capability secret -- we'll explain later why it's not
        // needed.
        None,
        director,
    )?;
```

Author: In many of the other code examples we've used literals and specific cases like "get everything by Sergio Leone". However, I think that's because it was bare code not wrapped in a function. I guess this *is* a function though -- happy to start wrapping more of this specific code in a function and making it generic if that seems better.

    match response {
        ZomeCallResponse::Ok(output) => output
            .decode()
            .map_err(|e| wasm_error!(e.to_string()))?,
        _ => Err(wasm_error!("Something bad happened")),
    }
}
```

This example calls that same function _if it's in a different cell_ whose role name is `movies`:

```rust
use hdk::prelude::*;
use movies_types::*;

> fn get_movies_by_sergio_leone_from_movies_cell() -> ExternResult<Vec<Movie>> {
>     let response = call(
>         // Address the cell by its role name.
>         // You can also address it by its cell ID.
>         CallTargetCell::OtherRole("movies".into()),
>         "movies",
>         "get_movies_for_director".into(),
>         None,
>         EntryHash::from_raw_36(vec![/* bytes of Sergio Leone director entry's hash */]),
>     )?;

Reviewer: ```suggestion
fn get_movies_for_directory_from_movies_cell(director: EntryHash) -> ExternResult<Vec<Movie>> {
    let response = call(
        // Address the cell by its role name.
        // You can also address it by its cell ID.
        CallTargetCell::OtherRole("movies".into()),
        "movies",
        "get_movies_for_director".into(),
        None,
        director,
    )?;
```

    match response {
        ZomeCallResponse::Ok(output) => output
            .decode()
            .map_err(|e| wasm_error!(e.to_string()))?,
        _ => Err(wasm_error!("Something bad happened")),
    }
}
```

> These cases don't need to worry about capability security either, because they're covered by a special grant called the [**author grant**](/concepts/8_calls_capabilities/#author-grant). It permits calls made by any caller with the same public key as the callee cell's owner.

Reviewer 2: I'm not sure this part is fully clear to me the way it's formulated. Is this saying that the call from zome A to zome B inside zome function X is allowed naturally as long the two zomes belong to the same cell (and thus same agent) and the call to zome function X is being made with a valid capability?

Author: Right again, and that applies to within the same cell _and_ to other cells within the same hApp. I'll try to explain the author grant a little more.

## Call a zome function from another agent in the network

If two agents have cells running the same DNA --- that is, they're part of the same network --- they can call each other's zome functions _in the same DNA_ using [`hdk::prelude::call_remote`](https://docs.rs/hdk/latest/hdk/p2p/fn.call_remote.html).

!!! info A remote cell might not be running the same coordinator zomes
Holochain allows agents to add and remove coordinator zomes from their cells. This permits upgrading and customization. But it also means that the zomes and functions that you _think_ are on the other end might not actually be there.
!!!

This example calls a function _in the same coordinator zome_ (or at least one with the same name) in a remote agent's cell. It assumes that the remote agent has granted access to their `get_movies_by_director` function with an [**unrestricted grant**](/concepts/8_calls_capabilities/#unrestricted), which doesn't require a capability secret.

```rust
use hdk::prelude::*;
use movies_types::*;

> fn get_movies_by_sergio_leone_remote() -> ExternResult<Vec<Movie>> {

Reviewer 2: potentially apply same changes here with entry hash as input as Jost suggested above for the other zome calls.

>     let bob_public_key = AgentPubKey::from_raw_36(vec![/* bytes of remote agent's key */]);

Reviewer 2: To be picky, I don't think that a remote agent's pubkey could be hard-coded that way here in practice because it would only be known at runtime so would need to be passed as a function argument. I'm not sure that that's relevant for the purpose of the docs though...

Reviewer 2: There's also the additional complexity in practice about actually getting that pubkey from somewhere. A typical pattern that works good for small DHTs at least is that every agent links their pubkey to an `ALL_AGENTS` anchor as part of `init()`.

Reviewer 2: But that's probably beyond the scope of this guide.

Author: :) I accidentally resolved this as part of resolving the suggestion to make the functions generic :)
